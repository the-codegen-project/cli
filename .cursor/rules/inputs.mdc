---
description: Input processing patterns and requirements for The Codegen Project CLI
globs: ["src/codegen/inputs/**/*.ts"]
alwaysApply: true
---

# Input Processing Rules

## Input Processing Pattern Compliance

### REQUIRED: Input processors must return standardized interfaces

```typescript
// Standard interface pattern for processed schema data
export interface Processed[Name]SchemaData {
  channelPayloads: Record<string, {schema: any; schemaId: string}>;
  operationPayloads: Record<string, {schema: any; schemaId: string}>;
  otherPayloads: {schema: any; schemaId: string}[];
}

// Alternative pattern for non-payload processors
export interface Processed[Name]Data {
  channel[Items]: Record<string, {schema: any; schemaId: string} | undefined>;
  operation[Items]: Record<string, {schema: any; schemaId: string} | undefined>;
  other[Items]: {schema: any; schemaId: string}[];
}
```

### Input Validation Pattern
```typescript
// REQUIRED: Input validation pattern
export function process[InputType][Name](document: [DocumentType]): Processed[Name]SchemaData {
  if (!document) {
    throw new Error('Expected [InputType] input, was not given');
  }
  
  // Processing logic here
  
  return {
    channelPayloads: {},
    operationPayloads: {},
    otherPayloads: []
  };
}
```

## Input Processing Patterns

### Separation of Concerns
- **Input Processors**: Located in `src/codegen/inputs/[input-type]/generators/`
- **Schema Extraction**: Extract schemas and return standardized `ProcessedXSchemaData` interfaces
- **Input Agnostic**: Core generators work with processed data, not raw input documents
- **Error Handling**: Validate inputs exist and throw descriptive errors

### File Organization
```
src/codegen/inputs/
├── asyncapi/
│   ├── parser.ts                    # Document parsing
│   ├── generators/
│   │   ├── payloads.ts             # Payload processing
│   │   ├── parameters.ts           # Parameter processing
│   │   ├── headers.ts              # Header processing
│   │   ├── types.ts                # Type processing
│   │   └── index.ts                # Barrel exports
│   └── index.ts                    # Main exports
├── openapi/
│   ├── parser.ts                   # Document parsing
│   ├── generators/
│   │   ├── payloads.ts             # Payload processing
│   │   ├── parameters.ts           # Parameter processing
│   │   ├── headers.ts              # Header processing
│   │   ├── types.ts                # Type processing
│   │   └── index.ts                # Barrel exports
│   └── index.ts                    # Main exports
```

## AsyncAPI Input Processing

### Document Parsing
```typescript
import {AsyncAPIDocumentInterface} from '@asyncapi/parser';

// REQUIRED: AsyncAPI document validation
export function processAsyncAPI[Name](
  asyncapiDocument: AsyncAPIDocumentInterface
): Processed[Name]SchemaData {
  if (!asyncapiDocument) {
    throw new Error('Expected AsyncAPI input, was not given');
  }

  // Extract channels, operations, and components
  const channels = asyncapiDocument.channels();
  const operations = asyncapiDocument.operations();
  const components = asyncapiDocument.components();

  // Process each category
  return {
    channelPayloads: processChannels(channels),
    operationPayloads: processOperations(operations),
    otherPayloads: processComponents(components)
  };
}
```

### Schema Extraction Patterns
```typescript
// Extract schema from AsyncAPI message
function extractMessageSchema(message: any): {schema: any; schemaId: string} | null {
  if (!message || !message.payload()) {
    return null;
  }

  const payload = message.payload();
  const schemaId = message.id() || generateSchemaId(message);
  
  return {
    schema: payload.json(),
    schemaId: pascalCase(schemaId)
  };
}

// Extract schema from AsyncAPI parameter
function extractParameterSchema(parameter: any): {schema: any; schemaId: string} | null {
  if (!parameter || !parameter.schema()) {
    return null;
  }

  return {
    schema: parameter.schema().json(),
    schemaId: pascalCase(parameter.id())
  };
}
```

## OpenAPI Input Processing

### Document Type Support
- **OpenAPI 2.0**: Support Swagger 2.0 specifications
- **OpenAPI 3.0**: Support OpenAPI 3.0 specifications  
- **OpenAPI 3.1**: Support OpenAPI 3.1 specifications with proper type guards

### Type Guards
```typescript
import {OpenAPIV2, OpenAPIV3, OpenAPIV3_1} from 'openapi-types';

// REQUIRED: OpenAPI document validation
export function processOpenAPI[Name](
  openapiDocument: OpenAPIV3.Document | OpenAPIV2.Document | OpenAPIV3_1.Document
): Processed[Name]SchemaData {
  if (!openapiDocument) {
    throw new Error('Expected OpenAPI input, was not given');
  }

  // Version-specific processing
  if (isOpenAPIV2(openapiDocument)) {
    return processOpenAPIV2[Name](openapiDocument);
  } else if (isOpenAPIV3(openapiDocument)) {
    return processOpenAPIV3[Name](openapiDocument);
  } else {
    return processOpenAPIV31[Name](openapiDocument);
  }
}
```

### Schema Extraction from Operations
```typescript
// Extract payload schemas from OpenAPI operations
function extractPayloadsFromOperations(
  paths: OpenAPIV3.PathsObject | OpenAPIV2.PathsObject | OpenAPIV3_1.PathsObject
): {
  requestPayloads: Record<string, {schema: any; schemaId: string}>;
  responsePayloads: Record<string, {schema: any; schemaId: string}>;
} {
  const requestPayloads: Record<string, {schema: any; schemaId: string}> = {};
  const responsePayloads: Record<string, {schema: any; schemaId: string}> = {};

  for (const [pathKey, pathItem] of Object.entries(paths)) {
    if (!pathItem) continue;

    for (const [method, operation] of Object.entries(pathItem)) {
      if (!operation || typeof operation !== 'object') continue;

      const operationId = operation.operationId ?? 
        `${method}${pathKey.replace(/[^a-zA-Z0-9]/g, '')}`;

      // Extract request schema
      const requestSchema = extractRequestSchema(operation);
      if (requestSchema) {
        requestPayloads[`${operationId}Request`] = requestSchema;
      }

      // Extract response schemas
      const responseSchemas = extractResponseSchemas(operation, operationId);
      Object.assign(responsePayloads, responseSchemas);
    }
  }

  return {requestPayloads, responsePayloads};
}
```

## Schema Processing Utilities

### Schema ID Generation
```typescript
// Generate consistent schema IDs
function generateSchemaId(source: any, fallback?: string): string {
  // Try to get ID from various sources
  const id = source.id?.() || 
             source.name?.() || 
             source.title ||
             source['x-modelina-name'] ||
             fallback ||
             'GeneratedSchema';
             
  return pascalCase(id);
}
```

### Schema Validation
```typescript
// Validate schema before processing
function validateSchema(schema: any, context: string): boolean {
  if (!schema) {
    Logger.warn(`No schema found for ${context}, skipping`);
    return false;
  }

  if (typeof schema !== 'object') {
    Logger.warn(`Invalid schema type for ${context}, expected object`);
    return false;
  }

  return true;
}
```

### Union Schema Creation
```typescript
// Create union schema from multiple schemas
function createUnionSchema(
  schemas: any[],
  baseId: string,
  hasStatusCodes: boolean = false
): any {
  if (schemas.length === 0) return null;
  if (schemas.length === 1) return schemas[0];

  const unionSchema = {
    $schema: 'http://json-schema.org/draft-07/schema',
    $id: baseId,
    title: baseId,
    oneOf: schemas,
    'x-modelina-has-status-codes': hasStatusCodes
  };

  return unionSchema;
}
```

## Error Handling Requirements

### Input Document Validation
```typescript
// REQUIRED: Always validate input documents exist
if (!document) {
  throw new Error('Expected [InputType] input, was not given');
}

// REQUIRED: Validate document structure
if (!document.channels && !document.paths) {
  throw new Error('Invalid document structure: missing channels or paths');
}
```

### Schema Processing Errors
```typescript
// Handle missing schemas gracefully
if (!schema) {
  Logger.warn(`No schema found for ${itemName}, skipping generation`);
  continue;
}

// Handle invalid schema types
if (typeof schema !== 'object') {
  Logger.error(`Invalid schema type for ${itemName}: ${typeof schema}`);
  throw new Error(`Cannot process non-object schema for ${itemName}`);
}
```

### Context Information in Errors
- Include document type (AsyncAPI/OpenAPI) in error messages
- Include channel/operation names when available
- Include schema paths for debugging
- Use descriptive error messages that help users fix issues

## Performance Considerations

### Memory Management
- Don't load entire documents into memory unnecessarily
- Process schemas incrementally when possible
- Use streaming for large document processing
- Cache parsed schemas when processing multiple generators

### Async Processing
```typescript
// REQUIRED: Use async processing for I/O operations
export async function processAsyncAPI[Name](
  asyncapiDocument: AsyncAPIDocumentInterface
): Promise<Processed[Name]SchemaData> {
  // Async processing logic
}
```

## Utility Functions

### Schema Transformation
```typescript
// Transform schema for Modelina compatibility
function transformSchemaForModelina(schema: any): any {
  // Add JSON Schema draft if missing
  if (!schema.$schema) {
    schema.$schema = 'http://json-schema.org/draft-07/schema';
  }

  // Ensure proper ID format
  if (schema.$id && !schema.$id.startsWith('#')) {
    schema.$id = `#/${schema.$id}`;
  }

  return schema;
}
```

### Component Processing
```typescript
// Extract schemas from document components
function extractComponentSchemas(
  components: any
): {schema: any; schemaId: string}[] {
  const componentSchemas: {schema: any; schemaId: string}[] = [];

  if (components?.schemas) {
    for (const [name, schema] of Object.entries(components.schemas)) {
      if (validateSchema(schema, `component.${name}`)) {
        componentSchemas.push({
          schema: transformSchemaForModelina(schema),
          schemaId: pascalCase(name)
        });
      }
    }
  }

  return componentSchemas;
}
```

## Required Imports
```typescript
// REQUIRED: Input processing imports
import {AsyncAPIDocumentInterface} from '@asyncapi/parser';
import {OpenAPIV2, OpenAPIV3, OpenAPIV3_1} from 'openapi-types';
import {Logger} from '../../../LoggingInterface';
import {pascalCase} from '../../generators/typescript/utils';
import {onlyUnique} from '../../utils';
```

## Forbidden Patterns in Input Processing
- ❌ **No direct document manipulation** - work with copies
- ❌ **No synchronous file operations** - use async alternatives
- ❌ **No hardcoded schema paths** - use dynamic extraction
- ❌ **No global state** - pass context through parameters
- ❌ **No `any` types** without proper validation
