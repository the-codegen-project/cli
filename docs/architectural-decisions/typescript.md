# Architectural decisions for TypeScript
All decision worth mentioning about TypeScript can be found here, simply to keep track.

### 21.08.2024

1. All functions in channel protocols are to be designed arrow function as it's more versatile when combining it together with other code sections.
