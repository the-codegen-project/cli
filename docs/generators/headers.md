---
sidebar_position: 99
---

# Headers

```js
export default {
  ...,
  generators: [
    {
      preset: 'headers',
      outputPath: './src/headers',
      serializationType: 'json', 
      language: 'typescript',
    }
  ]
};
```

`headers` preset is for generating models that represent typed models representing headers.

This is supported through the following inputs: [`asyncapi`](#inputs), [`openapi`](#inputs)

It supports the following languages; `typescript`

## Inputs

### `asyncapi`
The `headers` preset with `asyncapi` input generates all the message headers for each channel in the AsyncAPI document.

The return type is a map of channels and the model that represent the headers. 

### `openapi`
The `headers` preset with `openapi` input generates all the headers for each path in the OpenAPI document.

The return type is a map of paths and the model that represent the headers. 
