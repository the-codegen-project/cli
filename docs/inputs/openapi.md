---
sidebar_position: 99
---

# OpenAPI

Input support; `openapi`

- OpenAPI 3.0.x
- OpenAPI 3.1.x
- OpenAPI 2.0.0 (Swagger)

| **Presets** | OpenAPI | 
|---|---|
| [`payloads`](../generators/payloads.md) | ✅ |
| [`parameters`](../generators/parameters.md) | ✅ |
| [`headers`](../generators/headers.md) | ✅ |
| [`types`](../generators/types.md) | ✅ |
| [`channels`](../generators/channels.md) | ❌ |
| [`client`](../generators/client.md) | ❌ |
| [`custom`](../generators/custom.md) | ✅ |

## Basic Usage

### Configuration

Create a configuration file that specifies OpenAPI as the input type:

```json
{
  "inputType": "openapi",
  "inputPath": "./api/openapi.yaml",
  "language": "typescript",
  "generators": [ ... ]
}
```

## Advanced Features

### External References

The OpenAPI parser automatically resolves external `$ref` references:

```yaml
components:
  schemas:
    Pet:
      $ref: './schemas/pet.yaml#/Pet'
    User:
      $ref: 'https://api.example.com/schemas/user.json#/User'
```

### OpenAPI 3.1 Features

Full support for OpenAPI 3.1 features including:

- JSON Schema 2020-12 compatibility
- `const` keyword
- `if`/`then`/`else` conditionals
- Enhanced `examples` support

### Validation and Error Handling

The parser provides detailed validation errors:

```typescript
// If validation fails, you'll get detailed error information
try {
  const document = await loadOpenapi(context);
} catch (error) {
  console.error('OpenAPI validation failed:', error.message);
  // Error message includes line numbers and specific validation issues
}
```

## Examples

### REST API Client Generation

Generate a complete TypeScript client for your REST API:

```json
{
  "inputType": "openapi",
  "inputPath": "./api/openapi.yaml",
  "language": "typescript",
  "generators": [ ]
}
```

## Best Practices

1. **Schema Organization**: Use `$ref` to organize complex schemas into separate files
2. **Validation**: Always validate your OpenAPI documents before generation
3. **Versioning**: Include version information in your API specifications
4. **Documentation**: Use `description` fields extensively for better generated code
5. **Examples**: Provide examples in your schemas for better understanding

## Troubleshooting

### Common Issues

1. **Invalid $ref**: Ensure all referenced files exist and are accessible
2. **Schema Validation**: Check that your OpenAPI document follows the specification
3. **File Format**: Verify that YAML/JSON syntax is correct
4. **Circular References**: Avoid circular `$ref` dependencies

## FAQ

### Can I use both OpenAPI and AsyncAPI in the same project?

Yes! You can have separate configuration files for each input type and generate code to different output directories.

### What's the difference between OpenAPI 3.0 and 3.1?

OpenAPI 3.1 is fully compatible with JSON Schema 2020-12 and includes additional features like `const`, conditional schemas, and enhanced examples support.

### How do I handle authentication in generated clients?

Define security schemes in your OpenAPI document, and the generated client code will include appropriate authentication handling.

### Can I customize the generated code?

Yes, use the custom generator preset to create your own templates and generation logic. 