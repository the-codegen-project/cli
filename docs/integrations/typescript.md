---
sidebar_position: 99
---

# TypeScript Project

Simple TypeScript library that use AsyncAPI as input to generate payload models that is serialized and printed in the console.

[See the project in action ðŸŽ¬](https://github.com/the-codegen-project/cli/tree/main/examples/typescript-library)