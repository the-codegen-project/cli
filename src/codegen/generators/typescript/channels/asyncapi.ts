import {ChannelFunctionTypes} from './types';

type Action = 'send' | 'receive' | 'subscribe' | 'publish';
const sendingFunctionTypes = [
  ChannelFunctionTypes.NATS_JETSTREAM_PUBLISH,
  ChannelFunctionTypes.NATS_PUBLISH,
  ChannelFunctionTypes.NATS_REQUEST,
  ChannelFunctionTypes.MQTT_PUBLISH,
  ChannelFunctionTypes.KAFKA_PUBLISH,
  ChannelFunctionTypes.AMQP_EXCHANGE_PUBLISH,
  ChannelFunctionTypes.AMQP_QUEUE_PUBLISH,
  ChannelFunctionTypes.EVENT_SOURCE_EXPRESS
];
const receivingFunctionTypes = [
  ChannelFunctionTypes.NATS_JETSTREAM_PULL_SUBSCRIBE,
  ChannelFunctionTypes.NATS_JETSTREAM_PUSH_SUBSCRIBE,
  ChannelFunctionTypes.NATS_REPLY,
  ChannelFunctionTypes.NATS_SUBSCRIBE,
  ChannelFunctionTypes.KAFKA_SUBSCRIBE,
  ChannelFunctionTypes.EVENT_SOURCE_FETCH
];

// eslint-disable-next-line sonarjs/cognitive-complexity
export function shouldRenderFunctionType(
  givenFunctionTypes: ChannelFunctionTypes[] | undefined,
  functionTypesToCheckFor: ChannelFunctionTypes | ChannelFunctionTypes[],
  action: Action,
  reverseOperation: boolean
) {
  const listToCheck = [
    ...(Array.isArray(functionTypesToCheckFor)
      ? functionTypesToCheckFor
      : [functionTypesToCheckFor])
  ];
  const hasSendingOperation = action === 'send' || action === 'subscribe';
  const hasReceivingOperation = action === 'receive' || action === 'publish';
  const hasFunctionMappingConfig = givenFunctionTypes !== undefined;
  const checkForSending = listToCheck.some((item) =>
    sendingFunctionTypes.includes(item)
  );
  const checkForReceiving = listToCheck.some((item) =>
    receivingFunctionTypes.includes(item)
  );
  const hasFunctionType = (givenFunctionTypes ?? []).some((item) =>
    listToCheck.includes(item)
  );
  if (hasFunctionMappingConfig) {
    if (hasFunctionType) {
      const renderForSending = checkForSending && hasSendingOperation;
      const renderForReceiving = checkForReceiving && hasReceivingOperation;
      return renderForSending || renderForReceiving;
    }
    return false;
  }

  if (reverseOperation) {
    const renderForSending = hasSendingOperation && checkForReceiving;
    const renderForReceiving = hasReceivingOperation && checkForSending;
    return renderForSending || renderForReceiving;
  }

  const renderForSending = checkForSending && hasSendingOperation;
  const renderForReceiving = checkForReceiving && hasReceivingOperation;
  return renderForSending || renderForReceiving;
}
