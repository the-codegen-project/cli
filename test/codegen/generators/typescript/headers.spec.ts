import path from "node:path";
import { loadAsyncapiDocument } from "../../../../src/codegen/inputs/asyncapi";
import { generateTypescriptHeaders } from "../../../../src/codegen/generators/typescript/headers";
import { loadOpenapiDocument } from "../../../../src/codegen/inputs/openapi";

describe('headers', () => {
  describe('typescript', () => {
    it('should work with basic AsyncAPI inputs', async () => {
      const parsedAsyncAPIDocument = await loadAsyncapiDocument(path.resolve(__dirname, '../../../configs/headers.yaml'));
      
      const renderedContent = await generateTypescriptHeaders({
        generator: {
          outputPath: path.resolve(__dirname, './output'),
          preset: 'headers',
          language: 'typescript',
          dependencies: [],
          serializationType: 'json',
          id: 'test',
          includeValidation: true
        },
        inputType: 'asyncapi',
        asyncapiDocument: parsedAsyncAPIDocument,
        dependencyOutputs: { }
      });
      expect(renderedContent.channelModels['simple']!.result).toMatchSnapshot();
    });
    it('should work with OpenAPI 2.0 inputs', async () => {
      const parsedOpenAPIDocument = await loadOpenapiDocument(path.resolve(__dirname, '../../../configs/openapi-2.json'));
      
      const renderedContent = await generateTypescriptHeaders({
        generator: {
          outputPath: path.resolve(__dirname, './output'),
          preset: 'headers',
          language: 'typescript',
          dependencies: [],
          serializationType: 'json',
          includeValidation: true,
          id: 'test'
        },
        inputType: 'openapi',
        openapiDocument: parsedOpenAPIDocument,
        dependencyOutputs: { }
      });
      expect(renderedContent.channelModels['deletePet']!.result).toMatchSnapshot();
    });
    it('should work with OpenAPI 3.0 inputs', async () => {
      const parsedOpenAPIDocument = await loadOpenapiDocument(path.resolve(__dirname, '../../../configs/openapi-3.json'));
      
      const renderedContent = await generateTypescriptHeaders({
        generator: {
          outputPath: path.resolve(__dirname, './output'),
          preset: 'headers',
          language: 'typescript',
          dependencies: [],
          serializationType: 'json',
          id: 'test',
          includeValidation: true
        },
        inputType: 'openapi',
        openapiDocument: parsedOpenAPIDocument,
        dependencyOutputs: { }
      });
      expect(renderedContent.channelModels['deletePet']!.result).toMatchSnapshot();
    });
    it('should work with OpenAPI 3.1 inputs', async () => {
      const parsedOpenAPIDocument = await loadOpenapiDocument(path.resolve(__dirname, '../../../configs/openapi-3.json'));
      
      const renderedContent = await generateTypescriptHeaders({
        generator: {
          outputPath: path.resolve(__dirname, './output'),
          preset: 'headers',
          language: 'typescript',
          dependencies: [],
          serializationType: 'json',
          includeValidation: true,
          id: 'test'
        },
        inputType: 'openapi',
        openapiDocument: parsedOpenAPIDocument,
        dependencyOutputs: { }
      });
      expect(renderedContent.channelModels['deletePet']!.result).toMatchSnapshot();
    });
  });
});
