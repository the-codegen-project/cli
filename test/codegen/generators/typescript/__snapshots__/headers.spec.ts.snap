// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`headers typescript should work with OpenAPI 2.0 inputs 1`] = `
"import {Ajv, Options as AjvOptions, ErrorObject, ValidateFunction} from 'ajv';
import addFormats from 'ajv-formats';
class DeletePetHeaders {
  private _apiKey?: string;

  constructor(input: {
    apiKey?: string,
  }) {
    this._apiKey = input.apiKey;
  }

  get apiKey(): string | undefined { return this._apiKey; }
  set apiKey(apiKey: string | undefined) { this._apiKey = apiKey; }

  public marshal() : string {
    let json = '{'
    if(this.apiKey !== undefined) {
      json += \`"api_key": \${typeof this.apiKey === 'number' || typeof this.apiKey === 'boolean' ? this.apiKey : JSON.stringify(this.apiKey)},\`;
    }
  
    //Remove potential last comma 
    return \`\${json.charAt(json.length-1) === ',' ? json.slice(0, json.length-1) : json}}\`;
  }

  public static unmarshal(json: string | object): DeletePetHeaders {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new DeletePetHeaders({} as any);

    if (obj["api_key"] !== undefined) {
      instance.apiKey = obj["api_key"];
    }
  
  
    return instance;
  }
  public static theCodeGenSchema = {"type":"object","additionalProperties":false,"properties":{"api_key":{"type":"string"}},"$id":"DeletePetHeaders","$schema":"http://json-schema.org/draft-07/schema"};
  public static validate(context?: {data: any, ajvValidatorFunction?: ValidateFunction, ajvInstance?: Ajv, ajvOptions?: AjvOptions}): { valid: boolean; errors?: ErrorObject[]; } {
    const {data, ajvValidatorFunction} = context ?? {};
    const parsedData = typeof data === 'string' ? JSON.parse(data) : data;
    const validate = ajvValidatorFunction ?? this.createValidator(context)
    return {
      valid: validate(parsedData),
      errors: validate.errors ?? undefined,
    };
  }
  public static createValidator(context?: {ajvInstance?: Ajv, ajvOptions?: AjvOptions}): ValidateFunction {
    const {ajvInstance} = {...context ?? {}, ajvInstance: new Ajv(context?.ajvOptions ?? {})};
    addFormats(ajvInstance);
    ajvInstance.addVocabulary(["xml", "example"])
    const validate = ajvInstance.compile(this.theCodeGenSchema);
    return validate;
  }

}
export { DeletePetHeaders };"
`;

exports[`headers typescript should work with OpenAPI 3.0 inputs 1`] = `
"import {Ajv, Options as AjvOptions, ErrorObject, ValidateFunction} from 'ajv';
import addFormats from 'ajv-formats';
class DeletePetHeaders {
  private _apiKey?: string;

  constructor(input: {
    apiKey?: string,
  }) {
    this._apiKey = input.apiKey;
  }

  get apiKey(): string | undefined { return this._apiKey; }
  set apiKey(apiKey: string | undefined) { this._apiKey = apiKey; }

  public marshal() : string {
    let json = '{'
    if(this.apiKey !== undefined) {
      json += \`"api_key": \${typeof this.apiKey === 'number' || typeof this.apiKey === 'boolean' ? this.apiKey : JSON.stringify(this.apiKey)},\`;
    }
  
    //Remove potential last comma 
    return \`\${json.charAt(json.length-1) === ',' ? json.slice(0, json.length-1) : json}}\`;
  }

  public static unmarshal(json: string | object): DeletePetHeaders {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new DeletePetHeaders({} as any);

    if (obj["api_key"] !== undefined) {
      instance.apiKey = obj["api_key"];
    }
  
  
    return instance;
  }
  public static theCodeGenSchema = {"type":"object","additionalProperties":false,"properties":{"api_key":{"type":"string"}},"$id":"DeletePetHeaders","$schema":"http://json-schema.org/draft-07/schema"};
  public static validate(context?: {data: any, ajvValidatorFunction?: ValidateFunction, ajvInstance?: Ajv, ajvOptions?: AjvOptions}): { valid: boolean; errors?: ErrorObject[]; } {
    const {data, ajvValidatorFunction} = context ?? {};
    const parsedData = typeof data === 'string' ? JSON.parse(data) : data;
    const validate = ajvValidatorFunction ?? this.createValidator(context)
    return {
      valid: validate(parsedData),
      errors: validate.errors ?? undefined,
    };
  }
  public static createValidator(context?: {ajvInstance?: Ajv, ajvOptions?: AjvOptions}): ValidateFunction {
    const {ajvInstance} = {...context ?? {}, ajvInstance: new Ajv(context?.ajvOptions ?? {})};
    addFormats(ajvInstance);
    ajvInstance.addVocabulary(["xml", "example"])
    const validate = ajvInstance.compile(this.theCodeGenSchema);
    return validate;
  }

}
export { DeletePetHeaders };"
`;

exports[`headers typescript should work with OpenAPI 3.1 inputs 1`] = `
"import {Ajv, Options as AjvOptions, ErrorObject, ValidateFunction} from 'ajv';
import addFormats from 'ajv-formats';
class DeletePetHeaders {
  private _apiKey?: string;

  constructor(input: {
    apiKey?: string,
  }) {
    this._apiKey = input.apiKey;
  }

  get apiKey(): string | undefined { return this._apiKey; }
  set apiKey(apiKey: string | undefined) { this._apiKey = apiKey; }

  public marshal() : string {
    let json = '{'
    if(this.apiKey !== undefined) {
      json += \`"api_key": \${typeof this.apiKey === 'number' || typeof this.apiKey === 'boolean' ? this.apiKey : JSON.stringify(this.apiKey)},\`;
    }
  
    //Remove potential last comma 
    return \`\${json.charAt(json.length-1) === ',' ? json.slice(0, json.length-1) : json}}\`;
  }

  public static unmarshal(json: string | object): DeletePetHeaders {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new DeletePetHeaders({} as any);

    if (obj["api_key"] !== undefined) {
      instance.apiKey = obj["api_key"];
    }
  
  
    return instance;
  }
  public static theCodeGenSchema = {"type":"object","additionalProperties":false,"properties":{"api_key":{"type":"string"}},"$id":"DeletePetHeaders","$schema":"http://json-schema.org/draft-07/schema"};
  public static validate(context?: {data: any, ajvValidatorFunction?: ValidateFunction, ajvInstance?: Ajv, ajvOptions?: AjvOptions}): { valid: boolean; errors?: ErrorObject[]; } {
    const {data, ajvValidatorFunction} = context ?? {};
    const parsedData = typeof data === 'string' ? JSON.parse(data) : data;
    const validate = ajvValidatorFunction ?? this.createValidator(context)
    return {
      valid: validate(parsedData),
      errors: validate.errors ?? undefined,
    };
  }
  public static createValidator(context?: {ajvInstance?: Ajv, ajvOptions?: AjvOptions}): ValidateFunction {
    const {ajvInstance} = {...context ?? {}, ajvInstance: new Ajv(context?.ajvOptions ?? {})};
    addFormats(ajvInstance);
    ajvInstance.addVocabulary(["xml", "example"])
    const validate = ajvInstance.compile(this.theCodeGenSchema);
    return validate;
  }

}
export { DeletePetHeaders };"
`;

exports[`headers typescript should work with basic AsyncAPI inputs 1`] = `
"import {Ajv, Options as AjvOptions, ErrorObject, ValidateFunction} from 'ajv';
import addFormats from 'ajv-formats';
class SimpleObjectHeaders {
  private _displayName?: string;
  private _email?: string;
  private _additionalProperties?: Map<string, any>;

  constructor(input: {
    displayName?: string,
    email?: string,
    additionalProperties?: Map<string, any>,
  }) {
    this._displayName = input.displayName;
    this._email = input.email;
    this._additionalProperties = input.additionalProperties;
  }

  /**
   * Name of the user
   */
  get displayName(): string | undefined { return this._displayName; }
  set displayName(displayName: string | undefined) { this._displayName = displayName; }

  /**
   * Email of the user
   */
  get email(): string | undefined { return this._email; }
  set email(email: string | undefined) { this._email = email; }

  get additionalProperties(): Map<string, any> | undefined { return this._additionalProperties; }
  set additionalProperties(additionalProperties: Map<string, any> | undefined) { this._additionalProperties = additionalProperties; }

  public marshal() : string {
    let json = '{'
    if(this.displayName !== undefined) {
      json += \`"displayName": \${typeof this.displayName === 'number' || typeof this.displayName === 'boolean' ? this.displayName : JSON.stringify(this.displayName)},\`;
    }
    if(this.email !== undefined) {
      json += \`"email": \${typeof this.email === 'number' || typeof this.email === 'boolean' ? this.email : JSON.stringify(this.email)},\`;
    }
    if(this.additionalProperties !== undefined) { 
      for (const [key, value] of this.additionalProperties.entries()) {
        //Only unwrap those that are not already a property in the JSON object
        if(["displayName","email","additionalProperties"].includes(String(key))) continue;
        json += \`"\${key}": \${typeof value === 'number' || typeof value === 'boolean' ? value : JSON.stringify(value)},\`;
      }
    }
    //Remove potential last comma 
    return \`\${json.charAt(json.length-1) === ',' ? json.slice(0, json.length-1) : json}}\`;
  }

  public static unmarshal(json: string | object): SimpleObjectHeaders {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new SimpleObjectHeaders({} as any);

    if (obj["displayName"] !== undefined) {
      instance.displayName = obj["displayName"];
    }
    if (obj["email"] !== undefined) {
      instance.email = obj["email"];
    }
  
    instance.additionalProperties = new Map();
    const propsToCheck = Object.entries(obj).filter((([key,]) => {return !["displayName","email","additionalProperties"].includes(key);}));
    for (const [key, value] of propsToCheck) {
      instance.additionalProperties.set(key, value as any);
    }
    return instance;
  }
  public static theCodeGenSchema = {"type":"object","properties":{"displayName":{"type":"string","description":"Name of the user"},"email":{"type":"string","format":"email","description":"Email of the user"}},"$id":"SimpleObjectHeaders","$schema":"http://json-schema.org/draft-07/schema"};
  public static validate(context?: {data: any, ajvValidatorFunction?: ValidateFunction, ajvInstance?: Ajv, ajvOptions?: AjvOptions}): { valid: boolean; errors?: ErrorObject[]; } {
    const {data, ajvValidatorFunction} = context ?? {};
    const parsedData = typeof data === 'string' ? JSON.parse(data) : data;
    const validate = ajvValidatorFunction ?? this.createValidator(context)
    return {
      valid: validate(parsedData),
      errors: validate.errors ?? undefined,
    };
  }
  public static createValidator(context?: {ajvInstance?: Ajv, ajvOptions?: AjvOptions}): ValidateFunction {
    const {ajvInstance} = {...context ?? {}, ajvInstance: new Ajv(context?.ajvOptions ?? {})};
    addFormats(ajvInstance);
  
    const validate = ajvInstance.compile(this.theCodeGenSchema);
    return validate;
  }

}
export { SimpleObjectHeaders };"
`;
